src/tpm.rs

```diff
@@ -572,40 +584,56 @@ pub fn run<'a>(
 
                 thread::sleep(RETRY_SLEEP);
             }
-            _ => break,
+
+            _ => break 'exec,
         }
     }
 
-    let return_output = output.stdout;
-    let return_code = output.status.code();
+    // preprocess execution result
+    let return_output = String::from_utf8(output.stdout).map_err(|e| {
+        Box::new(format!(
+            "Can't convert output to utf8 encoded String. Error {}.",
+            e,
+        ))
+    })?;
+
+    // preprocess execution status code
+    let return_code = match output.status.code() {
+        Some(c) => c,
+        None => {
+            return emsg("Execution status code is None.", None::<String>);
+        }
+    };
 
-    if let (Some(c), true) = (return_code, raise_on_error) {
-        if c != except_code {
-            error!(
+    // Execution return code checking
+    if return_code != expect_code {
```

Comment:
Given that we compare return codes here, I don't see a need to include `return_code` in the returned value; if the function returns successfully, that should imply the codes match. Whoever called `run` knows the expected return code anyway; they shouldn't need it again.

If the codes don't match, we should call it out in an error specifically mentioning that as the cause and including the code that was returned (in which case it would be useful to the caller).

Reply:
You are right. I was thinking about the same thing. It will be better if the return code and the cause associated with the return code are wrapped together in one Err(). Box<String> isn't a ultimately return type for us apparently.

I can make an error type for the run() with another commit if that is ok?